### PR TITLE
226

### PR DESCRIPTION
**Pull request by maxkazmsft**
Closes https://github.com/microsoft/seismic-deeplearning/issues/226

### CODE REVIEW

**Comment by dciborow** [comment]
#ado microsoft.ComponentGovernance (seismic-deeplearning)

File: docker/README.md [hunk 1]
```diff
@@ -27,14 +27,15 @@ sudo systemctl restart docker
...
+sudo docker build -t seismic-deeplearning .
 ```
 This process will take a few minutes to complete. 
 
 # Run the Docker image:
-Once the Docker image is built, you can run it anytime using the following command: 
+Once the Docker image is built, you can run it anytime using the following command:
 ```bash
-sudo docker run --rm -it -p 8888:8888 --gpus=all --shm-size 11G --mount type=bind,source=$PWD/hrnetv2_w48_imagenet_pretrained.pth,target=/home/username/models/hrnetv2_w48_imagenet_pretrained.pth seismic-deeplearning
+sudo nvidia-docker run --rm -it -p 9000:9000 -p 9001:9001 --shm-size 11G --mount type=bind,source=$PWD/hrnetv2_w48_imagenet_pretrained.pth,target=/home/username/models/hrnetv2_w48_imagenet_pretrained.pth seismic-deeplearning
```
Comment (yalaudah) [inline docker/README.md]: This doesn't follow nvidia-docker's usage pattern (see here: https://github.com/NVIDIA/nvidia-docker#usage) 

Please also see this note from their readme: "Note that with the release of Docker 19.03, usage of nvidia-docker2 packages are deprecated since NVIDIA GPUs are now natively supported as devices in the Docker runtime." 

I think I might have used a newer version of docker  than you, thats why it worked for me probably and not for you?

File: examples/interpretation/notebooks/utilities.py [hunk 2]
```diff
@@ -319,38 +346,73 @@ def download_pretrained_model(config):
...
+
+            elif (
+                model == "deconvnet"
+                and "skip" not in config.MODEL.NAME
+                and config.TRAIN.DEPTH == "none"
+            ):
+                url = "http://deepseismicsharedstore.blob.core.windows.net/master-public-models/dutchf3_deconvnet_patch_no_depth.pth"
+            elif model == "unet" and config.TRAIN.DEPTH == "section":
+                url = "http://deepseismicsharedstore.blob.core.windows.net/master-public-models/dutchf3_seresnetunet_patch_section_depth.pth"
+            else:
```
Comment (yalaudah) [inline examples/interpretation/notebooks/utilities.py]: You added the other models, sweet!

File: experiments/interpretation/dutchf3_patch/local/configs/patch_deconvnet_skip.yaml [hunk 3]
```diff
@@ -12,7 +12,8 @@ SEED: 2019
 
 DATASET:
   NUM_CLASSES: 6
-  ROOT: /mnt/dutchf3
+  #ROOT: /mnt/dutchf3
```
Comment (yalaudah) [inline experiments/interpretation/dutchf3_patch/local/configs/patch_deconvnet_skip.yaml]:
```suggestion

```
Can we delete the comment?

File: experiments/interpretation/dutchf3_patch/local/test.py [hunk 4]
```diff
@@ -99,6 +106,21 @@ def reset(self):
...
+
+def mask_to_disk(mask, fname):
+    """
+    write segmentation mask to disk using a particular colormap
+    """
+    Image.fromarray(cm.gist_earth(normalize(mask), bytes=True)).save(fname)
+
+
 def _transform_CHW_to_HWC(numpy_array):
     return np.moveaxis(numpy_array, 0, -1)
```
Comment (yalaudah) [inline experiments/interpretation/dutchf3_patch/local/test.py]: I think these should be in a `utils` file somewhere, i've seen `_transform_CHW_to_HWC()` copied in mutliple places. Doesn't have to go in this PR, but something we should keep in mind.

File: experiments/interpretation/dutchf3_patch/local/test.py [hunk 5]
```diff
@@ -268,6 +317,19 @@ def _evaluate_split(
         logger.info("Running in Debug/Test mode")
         test_loader = take(1, test_loader)
```
Comment (yalaudah) [inline experiments/interpretation/dutchf3_patch/local/test.py]: I think for debug mode its better to avoid `toolz.take` and use PyTorch's built-in `torch.utils.data.Subset` instead, what do you think? toolz.take caused me a lot of problems when i tried to use it..

File: experiments/interpretation/dutchf3_patch/local/test.py [hunk 6]
```diff
@@ -350,12 +416,16 @@ def test(*options, cfg=None, debug=False):
...
+        [Normalize(mean=(config.TRAIN.MEAN,), std=(config.TRAIN.STD,), max_pixel_value=1,)]
+    )
 
     patch_aug = Compose(
         [
             Resize(
-                config.TRAIN.AUGMENTATIONS.RESIZE.HEIGHT, config.TRAIN.AUGMENTATIONS.RESIZE.WIDTH, always_apply=True,
+                config.TRAIN.AUGMENTATIONS.RESIZE.HEIGHT,
+                config.TRAIN.AUGMENTATIONS.RESIZE.WIDTH,
+                always_apply=True,
```
Comment (yalaudah) [inline experiments/interpretation/dutchf3_patch/local/test.py]: Are we sure we want to resize the patches before scoring them? I'm not sure  myself, In the quick experiment I ran today on training with and without resizing and padding, the former was better (especially for smaller classes -- one nice outcome of adding class IoUs in Tensorboard)

File: experiments/interpretation/dutchf3_patch/local/train.py [hunk 7]
```diff
@@ -158,12 +152,16 @@ def run(*options, cfg=None, debug=False):
         augmentations=val_aug,
     )
```
Comment (yalaudah) [inline experiments/interpretation/dutchf3_patch/local/train.py]: Related to the comment above. We might want to keep the augmentation as it is. 
```suggestion
)
basic_aug = Compose(
        [
            Normalize(mean=(config.TRAIN.MEAN,), std=(config.TRAIN.STD,), max_pixel_value=1),
            PadIfNeeded(
                min_height=config.TRAIN.PATCH_SIZE,
                min_width=config.TRAIN.PATCH_SIZE,
                border_mode=0,
                always_apply=True,
                mask_value=255,
                value=0,
            ),
            Resize(
                config.TRAIN.AUGMENTATIONS.RESIZE.HEIGHT, config.TRAIN.AUGMENTATIONS.RESIZE.WIDTH, always_apply=True,
            ),
            PadIfNeeded(
                min_height=config.TRAIN.AUGMENTATIONS.PAD.HEIGHT,
                min_width=config.TRAIN.AUGMENTATIONS.PAD.WIDTH,
                border_mode=0,
                always_apply=True,
                mask_value=255,
            ),
        ]
    )
```

File: docker/README.md (hunk 1 above)
Comment (yalaudah) [reply docker/README.md]: We might need to ask the user to update the Docker version 19.03 or higher like George suggested

File: experiments/interpretation/dutchf3_patch/local/test.py (hunk 5 above)
Comment (maxkazmsft) [reply experiments/interpretation/dutchf3_patch/local/test.py]: discussed and resolved with data.Subset attempt (until it breaks - then revert back to toolz.take)

File: experiments/interpretation/dutchf3_patch/local/test.py (hunk 6 above)
Comment (maxkazmsft) [reply experiments/interpretation/dutchf3_patch/local/test.py]: discussed - added TODO for another PR

File: experiments/interpretation/dutchf3_patch/local/train.py (hunk 7 above)
Comment (maxkazmsft) [reply experiments/interpretation/dutchf3_patch/local/train.py]: discussed - reverting back to older code
